sample/src/main/java/com/metaplex/sample/PhantomLoginViewModel.kt

```diff
@@ -0,0 +1,203 @@
+package com.metaplex.sample
+
+import android.app.Application
+import android.content.Context.MODE_PRIVATE
+import android.content.SharedPreferences
+import android.net.Uri
+import android.util.Log
+import android.widget.Toast
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.MutableLiveData
+import com.google.gson.Gson
+import com.solana.networking.Network
+import com.solana.vendor.TweetNaclFast
+import org.bitcoinj.core.Base58
+import java.net.URL
+import java.net.URLEncoder
+
+private const val PHANTOM_URL = "phantom.app"
+typealias Cluster = Network
+
+sealed class Result<out Success, out Failure> {
+    data class Success<out Success>(val value: Success) : Result<Success, Nothing>()
+    data class Failure<out Failure>(val reason: Failure) : Result<Nothing, Failure>()
+}
+
+data class DappKeyPair(val publicKey: ByteArray, val secretKey: ByteArray)
+
+interface PhantomDecryptor {
+    fun decryptPayload(data: String, nonce: String, sharedSecretDapp: ByteArray) : Result<ByteArray, PhantomError>
+}
+
+class NaCLDecryptor : PhantomDecryptor {
+    override fun decryptPayload(data: String, nonce: String, sharedSecretDapp: ByteArray) : Result<ByteArray, PhantomError> {
+        return try {
+//            val box = TweetNaclFast.Box(phantomEncryptionPublicKey, secretKey)
+//            val decryptedData = box.open(Base58.decode(data), Base58.decode(nonce))
+
+            val sbox = TweetNaclFast.SecretBox(sharedSecretDapp)
+            val decryptedData = sbox.open(Base58.decode(data), Base58.decode(nonce))
+
+            Result.Success(decryptedData)
+        } catch (err : Error) {
+            Result.Failure(PhantomError.CouldNotDecrypt(err))
+        }
+    }
+}
+
+data class PhantomErrorResponse(val errorCode: String, val errorMessage: String)
+
+sealed class PhantomError {
+    data class HostNotSupported(val host: String) : PhantomError()
+    data class SchemeNotFound(val scheme: String) : PhantomError()
+    data class ErrorResponse(val phantomErrorResponse: PhantomErrorResponse) : PhantomError()
+    data class CouldNotDecrypt(val error: Error) : PhantomError()
+    data class CouldNotGenerateSharedSecret(val error: Error) : PhantomError()
+    object InvalidParameters : PhantomError()
+}
+
+data class PhantomConnectResponseData(val public_key: String, val session: String)
+
+sealed class PhantomResponse {
+    data class OnConnect(val response: PhantomConnectResponseData, val sharedSecretDapp: String, val phantomEncryptionPublicKey: String) : PhantomResponse()
+}
+
+enum class PhantomAction {
+    connect {
+        override fun host() = "sampleonconnect"
+    };
+    abstract fun host() : String
+}
+
+
+class PhantomLoginViewModel(application: Application) : AndroidViewModel(application) {
+    private val context = getApplication<Application>().applicationContext
```

Comment:
@thedevyansh this is actually a memory leak: context leaking. The context should never be saved outside of the lifecycle. It can be passed in and references, but should not be saved. 

This probably wont be a big problem for this sample app, but this kind of leak is what causes issues like [this one](https://github.com/metaplex-foundation/metaplex-android/issues/49)

Reply:
@Funkatronics, I completely forgot to ask about this concern! I am familiar with this issue. I required the context to save the session via shared preference. I couldn't save it in the Login activity since the view model knows nothing about the activity by design. How else can I store the session?

Reply:
saving data via shared preference is fine (for a sample app). Using SharedPrefs is not the problem, saving a reference to context is the problem. The typical practice is to use the context during construction to get whatever you need, then throw away the context reference (don't save it in a local variable). This allows you to access the context without leaking it. 

Also, I recommend to use `ViewModel` instead of `AndroidViewModel`. It's not a big deal, but its an easy change so I think we should switch to `ViewModel`. I can explain in more detail about why this is better if you'd like.

Reply:
@Funkatronics Removed the context from viewmodel. Please review.